File: templates/.claude/hooks/scan-prompt-secrets.py
#!/usr/bin/env python3
"""
Claude Code UserPromptSubmit hook: scans prompts for leaked secrets before submission.
Blocks prompts that contain API keys, tokens, passwords, etc.
"""

import json
import re
import sys

# Secret patterns with names
SECRET_PATTERNS = [
    # API Keys
    (r"sk-[a-zA-Z0-9]{20,}", "OpenAI API key"),
    (r"sk-proj-[a-zA-Z0-9\-_]{40,}", "OpenAI project API key"),
    (r"sk-ant-[a-zA-Z0-9\-_]{40,}", "Anthropic API key"),
    (r"AIza[0-9A-Za-z\-_]{35}", "Google API key"),
    (r"AKIA[0-9A-Z]{16}", "AWS Access Key ID"),
    (r"ghp_[a-zA-Z0-9]{36}", "GitHub Personal Access Token"),
    (r"gho_[a-zA-Z0-9]{36}", "GitHub OAuth Token"),
    (r"github_pat_[a-zA-Z0-9]{22}_[a-zA-Z0-9]{59}", "GitHub PAT (fine-grained)"),
    (r"glpat-[a-zA-Z0-9\-_]{20,}", "GitLab Personal Access Token"),
    (r"xox[baprs]-[0-9]{10,13}-[0-9]{10,13}-[a-zA-Z0-9]{24}", "Slack Token"),
    (r"https://hooks\.slack\.com/services/T[a-zA-Z0-9]{8}/B[a-zA-Z0-9]{8,}/[a-zA-Z0-9]{24}", "Slack Webhook"),
    (r"sq0atp-[0-9A-Za-z\-_]{22}", "Square Access Token"),
    (r"sq0csp-[0-9A-Za-z\-_]{43}", "Square OAuth Secret"),
    (r"SG\.[a-zA-Z0-9]{22}\.[a-zA-Z0-9]{43}", "SendGrid API Key"),
    (r"key-[a-zA-Z0-9]{32}", "Mailgun API Key"),
    (r"[a-f0-9]{32}-us[0-9]{1,2}", "Mailchimp API Key"),
    (r"sk_live_[0-9a-zA-Z]{24,}", "Stripe Secret Key"),
    (r"rk_live_[0-9a-zA-Z]{24,}", "Stripe Restricted Key"),
    (r"AC[a-zA-Z0-9]{32}", "Twilio Account SID"),
    (r"np_[a-z0-9]{36}", "npm token"),
    (r"pypi-AgEIcHlwaS5vcmc[a-zA-Z0-9\-_]{50,}", "PyPI API Token"),
    # Azure (removed overly broad base64 pattern)
    (r"DefaultEndpointsProtocol=https;AccountName=[^;]+;AccountKey=[^;]+", "Azure Connection String"),
    # Database connection strings
    (r"postgres(ql)?://[^:]+:[^@]+@", "PostgreSQL connection string with password"),
    (r"mysql://[^:]+:[^@]+@", "MySQL connection string with password"),
    (r"mongodb(\+srv)?://[^:]+:[^@]+@", "MongoDB connection string with password"),
    (r"redis://:[^@]+@", "Redis connection string with password"),
    # Private keys
    (r"-----BEGIN (RSA |EC |DSA |OPENSSH )?PRIVATE KEY-----", "Private key"),
    (r"-----BEGIN PGP PRIVATE KEY BLOCK-----", "PGP private key"),
    # JWT with signature
    (r"eyJ[a-zA-Z0-9_-]*\.eyJ[a-zA-Z0-9_-]*\.[a-zA-Z0-9_-]+", "JWT token"),
    # Generic patterns (less specific, use with caution)
    (r"password\s*[=:]\s*['\"][^'\"]{8,}['\"]", "Hardcoded password"),
    (r"api[_-]?key\s*[=:]\s*['\"][a-zA-Z0-9]{16,}['\"]", "Generic API key"),
    (r"secret\s*[=:]\s*['\"][a-zA-Z0-9]{16,}['\"]", "Generic secret"),
]

# Compile patterns
COMPILED_PATTERNS = [(re.compile(p, re.IGNORECASE), name) for p, name in SECRET_PATTERNS]


def main() -> None:
    try:
        input_data = json.load(sys.stdin)
    except json.JSONDecodeError:
        sys.exit(0)

    prompt = input_data.get("prompt", "")

    if not prompt:
        sys.exit(0)

    found_secrets: list[str] = []

    for pattern, name in COMPILED_PATTERNS:
        if pattern.search(prompt):
            found_secrets.append(name)

    if found_secrets:
        unique_secrets = list(set(found_secrets))
        secret_list = "\n  - ".join(unique_secrets)
        print(
            f"Potential secrets detected in prompt:\n  - {secret_list}\n\nPlease remove sensitive data before submitting.",
            file=sys.stderr,
        )
        sys.exit(2)

    sys.exit(0)


if __name__ == "__main__":
    main()

File: templates/opencode/plugins/scan-prompt-secrets.ts
import type { Plugin } from "@opencode-ai/plugin"

/**
 * OpenCode plugin: scans prompts for leaked secrets before submission.
 * Blocks prompts that contain API keys, tokens, passwords, etc.
 */

// Secret patterns with names
const SECRET_PATTERNS: [RegExp, string][] = [
  // API Keys
  [/sk-[a-zA-Z0-9]{20,}/i, "OpenAI API key"],
  [/sk-proj-[a-zA-Z0-9\-_]{40,}/i, "OpenAI project API key"],
  [/sk-ant-[a-zA-Z0-9\-_]{40,}/i, "Anthropic API key"],
  [/AIza[0-9A-Za-z\-_]{35}/i, "Google API key"],
  [/AKIA[0-9A-Z]{16}/i, "AWS Access Key ID"],
  [/ghp_[a-zA-Z0-9]{30,40}/i, "GitHub Personal Access Token"],
  [/gho_[a-zA-Z0-9]{30,40}/i, "GitHub OAuth Token"],
  [/github_pat_[a-zA-Z0-9]{22}_[a-zA-Z0-9]{59}/i, "GitHub PAT (fine-grained)"],
  [/glpat-[a-zA-Z0-9\-_]{20,}/i, "GitLab Personal Access Token"],
  [/xox[baprs]-[0-9]{10,13}-[0-9]{10,13}-[a-zA-Z0-9]{24}/i, "Slack Token"],
  [/https:\/\/hooks\.slack\.com\/services\/T[a-zA-Z0-9]{8}\/B[a-zA-Z0-9]{8,}\/[a-zA-Z0-9]{24}/i, "Slack Webhook"],
  [/sq0atp-[0-9A-Za-z\-_]{22}/i, "Square Access Token"],
  [/sq0csp-[0-9A-Za-z\-_]{43}/i, "Square OAuth Secret"],
  [/SG\.[a-zA-Z0-9]{22}\.[a-zA-Z0-9]{43}/i, "SendGrid API Key"],
  [/key-[a-zA-Z0-9]{32}/i, "Mailgun API Key"],
  [/[a-f0-9]{32}-us[0-9]{1,2}/i, "Mailchimp API Key"],
  [/sk_live_[0-9a-zA-Z]{24,}/i, "Stripe Secret Key"],
  [/rk_live_[0-9a-zA-Z]{24,}/i, "Stripe Restricted Key"],
  [/AC[a-zA-Z0-9]{32}/i, "Twilio Account SID"],
  [/np_[a-z0-9]{36}/i, "npm token"],
  [/pypi-AgEIcHlwaS5vcmc[a-zA-Z0-9\-_]{50,}/i, "PyPI API Token"],
  // Azure
  [/DefaultEndpointsProtocol=https;AccountName=[^;]+;AccountKey=[^;]+/i, "Azure Connection String"],
  // Database connection strings
  [/postgres(ql)?:\/\/[^:]+:[^@]+@/i, "PostgreSQL connection string with password"],
  [/mysql:\/\/[^:]+:[^@]+@/i, "MySQL connection string with password"],
  [/mongodb(\+srv)?:\/\/[^:]+:[^@]+@/i, "MongoDB connection string with password"],
  [/redis:\/\/:[^@]+@/i, "Redis connection string with password"],
  // Private keys
  [/-----BEGIN (RSA |EC |DSA |OPENSSH )?PRIVATE KEY-----/i, "Private key"],
  [/-----BEGIN PGP PRIVATE KEY BLOCK-----/i, "PGP private key"],
  // JWT with signature
  [/eyJ[a-zA-Z0-9_-]*\.eyJ[a-zA-Z0-9_-]*\.[a-zA-Z0-9_-]+/i, "JWT token"],
  // Generic patterns (less specific, use with caution)
  [/password\s*[=:]\s*['"'][^'"']{8,}['"']/i, "Hardcoded password"],
  [/api[_-]?key\s*[=:]\s*['"'][a-zA-Z0-9]{16,}['"']/i, "Generic API key"],
  [/secret\s*[=:]\s*['"'][a-zA-Z0-9]{16,}['"']/i, "Generic secret"],
]

export const ScanPromptSecretsPlugin: Plugin = async () => {
  return {
    "tui.prompt.append": async (input, output) => {
      const prompt = output.prompt
      
      if (!prompt) {
        return
      }

      const foundSecrets: string[] = []

      for (const [pattern, name] of SECRET_PATTERNS) {
        if (pattern.test(prompt)) {
          foundSecrets.push(name)
        }
      }

      if (foundSecrets.length > 0) {
        const uniqueSecrets = [...new Set(foundSecrets)]
        const secretList = uniqueSecrets.map(s => `  - ${s}`).join("\n")
        
        throw new Error(
          `Potential secrets detected in prompt:\n${secretList}\n\nPlease remove sensitive data before submitting.`
        )
      }
    },
  }
}
